FlowAggregator e2e tests sometimes timed out.

**Describe the bug**
When we run `/test-ipv6-e2e` to trigger our CI job `jenkins-ipv6-ds-e2e`, the `TestFlowAggregator` sometimes times out. It's kind freaky. 
For example, http://10.176.27.169:8080/job/antrea-ipv6-ds-e2e-for-pull-request/279, this is triggered by PR #2034, which only fixes some small bugs in the TraceFlow test. But all tests under `TestFlowAggregator/` fails. 
The related log is below:
```
=== RUN   TestFlowAggregator/IPv4
=== RUN   TestFlowAggregator/IPv4/IntraNodeFlows
    flowaggregator_test.go:390: Network Policies are realized.
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:152
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv4/IntraNodeFlows
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== RUN   TestFlowAggregator/IPv4/InterNodeFlows
    flowaggregator_test.go:388: Error when waiting for Network Policy to be realized: unable to upgrade connection: container not found ("antrea-agent")
    flowaggregator_test.go:390: Network Policies are realized.
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:175
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv4/InterNodeFlows
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== RUN   TestFlowAggregator/IPv4/LocalServiceAccess
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:186
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv4/LocalServiceAccess
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== RUN   TestFlowAggregator/IPv4/RemoteServiceAccess
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:196
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv4/RemoteServiceAccess
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== RUN   TestFlowAggregator/IPv6
=== RUN   TestFlowAggregator/IPv6/IntraNodeFlows
    flowaggregator_test.go:390: Network Policies are realized.
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:154
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv6/IntraNodeFlows
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== RUN   TestFlowAggregator/IPv6/InterNodeFlows
    flowaggregator_test.go:390: Network Policies are realized.
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:177
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv6/InterNodeFlows
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== RUN   TestFlowAggregator/IPv6/LocalServiceAccess
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:188
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv6/LocalServiceAccess
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== RUN   TestFlowAggregator/IPv6/RemoteServiceAccess
    flowaggregator_test.go:226: 
        	Error Trace:	flowaggregator_test.go:226
        	            				flowaggregator_test.go:198
        	Error:      	Received unexpected error:
        	            	timed out waiting for the condition
        	Test:       	TestFlowAggregator/IPv6/RemoteServiceAccess
        	Messages:   	IPFIX collector did not receive the expected records and timed out with error: timed out waiting for the condition
=== CONT  TestFlowAggregator
    fixtures.go:319: Deleting 'flow-aggregator' K8s Namespace
    fixtures.go:208: Exporting test logs to '/var/lib/jenkins/workspace/antrea-ipv6-ds-e2e-for-pull-request/antrea-test-logs/TestFlowAggregator/beforeTeardown.Apr06-23-30-51'
    fixtures.go:309: Error when exporting kubelet logs: error when running journalctl on Node 'antrea-ipv6-7-0', is it available?
    fixtures.go:330: Deleting 'antrea-test' K8s Namespace
--- FAIL: TestFlowAggregator (409.90s)
    --- FAIL: TestFlowAggregator/IPv4 (105.62s)
        --- FAIL: TestFlowAggregator/IPv4/IntraNodeFlows (21.91s)
        --- FAIL: TestFlowAggregator/IPv4/InterNodeFlows (37.18s)
        --- FAIL: TestFlowAggregator/IPv4/LocalServiceAccess (21.29s)
        --- FAIL: TestFlowAggregator/IPv4/RemoteServiceAccess (22.15s)
    --- FAIL: TestFlowAggregator/IPv6 (153.10s)
        --- FAIL: TestFlowAggregator/IPv6/IntraNodeFlows (21.98s)
        --- FAIL: TestFlowAggregator/IPv6/InterNodeFlows (21.88s)
        --- FAIL: TestFlowAggregator/IPv6/LocalServiceAccess (84.81s)
        --- FAIL: TestFlowAggregator/IPv6/RemoteServiceAccess (21.33s)
```

Sure Zhecheng.
There are two aspects here. One is the failure itself and the timeout values when the failure happens.
1. These failures were observed on the dual-stack cluster in PR #1904 as well. Yiou fixed this in her PR. She observed that the test is passing consistently after the fix. Could you confirm this @zyiou?
2. Timeout values: `collectorCheckTimeout` value of 10s is high. This can be reduced to flowAggregator export interval value. This will reduce the typical 20s to 15s. However, the following values are very surprising:

>         --- FAIL: TestFlowAggregator/IPv4/InterNodeFlows (37.18s)
>         --- FAIL: TestFlowAggregator/IPv6/LocalServiceAccess (84.81s)

I could not reproduce this locally but I am suspecting `wait.Poll` condition takes a long time. There could be issues when grabbing the kubectl log output from`ipfix-collector'. Maybe the pod is in a messed up state, so added an option `--pod-running-timeout=xx`. 
Fixing the timeout values by opening a PR now.

> 1. These failures were observed on the dual-stack cluster in PR #1904 as well. Yiou fixed this in her PR. She observed that the test is passing consistently after the fix. Could you confirm this @zyiou?

The failure I had was kinda different: `wait.Poll` condition got passed but incorrect number of records were received (expecting 2 or more but only 1 was received). So I changed the wait condition to pass when 2 or more valid records are found. 

I found this error message might be the reason:
```
flowaggregator_test.go:388: Error when waiting for Network Policy to be realized: unable to upgrade connection: container not found ("antrea-agent")
```
Since all tests are failing, probably flow exporter does not start correctly.

Thanks, Yiou. This seems like an intermittent issue--I did not hit this when running tests as part of PR #2039. Tried running this multiple times with `count` argument locally as well. 

I see that previous test cases have passed but as you pointed out in the following error that the Antrea agent is missing, but the same error was not there when configuring network policies in the other three test cases, where they are required. 

> ```
> flowaggregator_test.go:388: Error when waiting for Network Policy to be realized: unable to upgrade connection: container not found ("antrea-agent")
> ```
> 

Focusing on timeout values, the above error may or may not incur [some extra delay](https://github.com/vmware-tanzu/antrea/blob/main/test/e2e/performance_test.go#L279) but the `TestFlowAggregator/IPv6/LocalServiceAccess (84.81s)` do not have any network policy configuration. Not sure where the extra time is coming here from.
With PR #2039 , we try to cut down any possible delay when waiting for flow records from the ipfix-collector pod logs.

I see some intermittent e2e test errors in other testsuites as well. I was able to get the logs on one of the runs and found the backtrace where flow aggregator crashes.
https://jenkins.antrea-ci.rocks/job/antrea-e2e-for-pull-request/2382/consoleFull

```
E0414 17:21:26.168044       6 aggregate.go:144] Callback execution failed for flow with key: {100.96.1.2 192.168.4.38 6 60160 10350}, records: {0xc0001461f8 true true}, error: error when sending data set: error when sending message on the connection: write tcp 100.96.1.23:54482->192.168.4.38:4739: write: connection reset by peer
E0414 17:21:26.168674       6 flowaggregator.go:352] Error when sending flow records: error when sending data set: error when sending message on the connection: write tcp 100.96.1.23:54482->192.168.4.38:4739: write: connection reset by peer
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x119a432]

goroutine 15 [running]:
github.com/vmware/go-ipfix/pkg/entities.(*dataRecord).GetMinDataRecordLen(0xc000010468, 0xc0000f3e00)
	<autogenerated>:1 +0x32
github.com/vmware/go-ipfix/pkg/exporter.(*ExportingProcess).SendSet(0xc0002a8040, 0x1673460, 0xc0002e76e0, 0x13a0100, 0x0, 0x0)
	/go/pkg/mod/github.com/vmware/go-ipfix@v0.4.7/pkg/exporter/process.go:168 +0x2a2
github.com/vmware-tanzu/antrea/pkg/ipfix.(*ipfixExportingProcess).SendSet(0xc000146160, 0x1667c60, 0xc0002ee4c0, 0x40, 0xc000440100, 0x0)
	/antrea/pkg/ipfix/ipfix_process.go:48 +0x48
github.com/vmware-tanzu/antrea/pkg/flowaggregator.(*flowAggregator).sendTemplateSet(0xc0001eb2c0, 0x1667c60, 0xc0002ee4c0, 0x0, 0x9a41db2e, 0x0, 0x0)
	/antrea/pkg/flowaggregator/flowaggregator.go:453 +0xa47
github.com/vmware-tanzu/antrea/pkg/flowaggregator.(*flowAggregator).initExportingProcess(0xc0001eb2c0, 0xc00006aef0, 0x2)
	/antrea/pkg/flowaggregator/flowaggregator.go:299 +0x355
github.com/vmware-tanzu/antrea/pkg/flowaggregator.(*flowAggregator).Run(0xc0001eb2c0, 0xc0001626c0)
	/antrea/pkg/flowaggregator/flowaggregator.go:343 +0x354
created by github.com/vmware-tanzu/antrea/cmd/flow-aggregator.run
	/antrea/cmd/flow-aggregator/flow-aggregator.go:107 +0x3de

```
After closely examining the backtrac, I found that records in `fa.set` are not reset properly. We use the same set for both template and data records. Following is the code of `set.Reset()` in go-ipfix.

```
func (s *set) ResetSet() {
	s.buffer.Reset()
	s.setType = Undefined
	s.records = s.records[:0]
}
```
I read in the go documentation and also on [stackoverflow](https://stackoverflow.com/questions/16971741/how-do-you-clear-a-slice-in-go#:~:text=Setting%20the%20slice%20to%20nil,memory%20to%20the%20garbage%20collector.&text=Note%20that%20slices%20can%20easily,to%20the%20same%20underlying%20memory.) `s.records = s.records[:0]` may not be picked by GC and the safest way to do this is by using `s.records = nil`. I will fix this in go-ipfix through a patch release and will upgrade the package version in Antrea.

--- a/pkg/agent/flowexporter/exporter/exporter.go
+++ b/pkg/agent/flowexporter/exporter/exporter.go
@@ -232,11 +232,7 @@ func (exp *flowExporter) initFlowExporter() error {
 	if exp.v4Enabled {
 		templateID := exp.process.NewTemplateID()
 		exp.templateIDv4 = templateID
-		if err = exp.ipfixSet.PrepareSet(ipfixentities.Template, exp.templateIDv4); err != nil {
-			return err
-		}
-		sentBytes, err := exp.sendTemplateSet(exp.ipfixSet, false)
-		exp.ipfixSet.ResetSet()
+		sentBytes, err := exp.sendTemplateSet(false)
 		if err != nil {
 			return err
 		}
@@ -246,11 +242,7 @@ func (exp *flowExporter) initFlowExporter() error {
 	if exp.v6Enabled {
 		templateID := exp.process.NewTemplateID()
 		exp.templateIDv6 = templateID
-		if err = exp.ipfixSet.PrepareSet(ipfixentities.Template, exp.templateIDv6); err != nil {
-			return err
-		}
-		sentBytes, err := exp.sendTemplateSet(exp.ipfixSet, true)
-		exp.ipfixSet.ResetSet()
+		sentBytes, err := exp.sendTemplateSet(true)
 		if err != nil {
 			return err
 		}
@@ -359,7 +351,7 @@ func (exp *flowExporter) sendFlowRecords() error {
 	return nil
 }
 
-func (exp *flowExporter) sendTemplateSet(templateSet ipfixentities.Set, isIPv6 bool) (int, error) {
+func (exp *flowExporter) sendTemplateSet(isIPv6 bool) (int, error) {
 	elements := make([]*ipfixentities.InfoElementWithValue, 0)
 
 	IANAInfoElements := IANAInfoElementsIPv4
@@ -394,13 +386,15 @@ func (exp *flowExporter) sendTemplateSet(templateSet ipfixentities.Set, isIPv6 b
 		ieWithValue := ipfixentities.NewInfoElementWithValue(element, nil)
 		elements = append(elements, ieWithValue)
 	}
-
-	err := templateSet.AddRecord(elements, templateID)
+	exp.ipfixSet.ResetSet()
+	if err := exp.ipfixSet.PrepareSet(ipfixentities.Template, templateID); err != nil {
+		return 0, err
+	}
+	err := exp.ipfixSet.AddRecord(elements, templateID)
 	if err != nil {
 		return 0, fmt.Errorf("error in adding record to template set: %v", err)
 	}
-
-	sentBytes, err := exp.process.SendSet(templateSet)
+	sentBytes, err := exp.process.SendSet(exp.ipfixSet)
 	if err != nil {
 		return 0, fmt.Errorf("error in IPFIX exporting process when sending template record: %v", err)
 	}

--- a/pkg/agent/flowexporter/exporter/exporter_test.go
+++ b/pkg/agent/flowexporter/exporter/exporter_test.go
@@ -96,7 +96,7 @@ func testSendTemplateSet(t *testing.T, v4Enabled bool, v6Enabled bool) {
 func sendTemplateSet(t *testing.T, ctrl *gomock.Controller, mockIPFIXExpProc *ipfixtest.MockIPFIXExportingProcess, mockIPFIXRegistry *ipfixtest.MockIPFIXRegistry, flowExp *flowExporter, isIPv6 bool) {
 	var mockTempSet *ipfixentitiestesting.MockSet
 	mockTempSet = ipfixentitiestesting.NewMockSet(ctrl)
-
+	flowExp.ipfixSet = mockTempSet
 	// Following consists of all elements that are in IANAInfoElements and AntreaInfoElements (globals)
 	// Only the element name is needed, other arguments have dummy values.
 	var elemList = make([]*ipfixentities.InfoElementWithValue, 0)
@@ -125,8 +125,14 @@ func sendTemplateSet(t *testing.T, ctrl *gomock.Controller, mockIPFIXExpProc *ip
 	}
 	// Passing 0 for sentBytes as it is not used anywhere in the test. If this not a call to mock, the actual sentBytes
 	// above elements: IANAInfoElements, IANAReverseInfoElements and AntreaInfoElements.
+	mockTempSet.EXPECT().ResetSet()
+	if !isIPv6 {
+		mockTempSet.EXPECT().PrepareSet(ipfixentities.Template, testTemplateIDv4).Return(nil)
+	} else {
+		mockTempSet.EXPECT().PrepareSet(ipfixentities.Template, testTemplateIDv6).Return(nil)
+	}
 	mockIPFIXExpProc.EXPECT().SendSet(mockTempSet).Return(0, nil)
-	_, err := flowExp.sendTemplateSet(mockTempSet, isIPv6)
+	_, err := flowExp.sendTemplateSet(isIPv6)
 	assert.NoError(t, err, "Error in sending template set")
 
 	eL := flowExp.elementsListv4

--- a/pkg/flowaggregator/flowaggregator.go
+++ b/pkg/flowaggregator/flowaggregator.go
@@ -317,10 +317,7 @@ func (fa *flowAggregator) initExportingProcess() error {
 	fa.exportingProcess = ep
 	// Currently, we send two templates for IPv4 and IPv6 regardless of the IP families supported by cluster
 	fa.templateIDv4 = fa.exportingProcess.NewTemplateID()
-	if err := fa.set.PrepareSet(ipfixentities.Template, fa.templateIDv4); err != nil {
-		return fmt.Errorf("error when preparing IPv4 template set: %v", err)
-	}
-	bytesSent, err := fa.sendTemplateSet(fa.set, false)
+	bytesSent, err := fa.sendTemplateSet(false)
 	if err != nil {
 		fa.exportingProcess.CloseConnToCollector()
 		fa.exportingProcess = nil
@@ -330,12 +327,8 @@ func (fa *flowAggregator) initExportingProcess() error {
 	}
 	klog.V(2).Infof("Initialized exporting process and sent %d bytes size of IPv4 template set", bytesSent)
 
-	fa.set.ResetSet()
 	fa.templateIDv6 = fa.exportingProcess.NewTemplateID()
-	if err := fa.set.PrepareSet(ipfixentities.Template, fa.templateIDv6); err != nil {
-		return fmt.Errorf("error when preparing IPv6 template set: %v", err)
-	}
-	bytesSent, err = fa.sendTemplateSet(fa.set, true)
+	bytesSent, err = fa.sendTemplateSet(true)
 	if err != nil {
 		fa.exportingProcess.CloseConnToCollector()
 		fa.exportingProcess = nil
@@ -426,7 +419,7 @@ func (fa *flowAggregator) sendFlowKeyRecord(key ipfixintermediate.FlowKey, recor
 	return nil
 }
 
-func (fa *flowAggregator) sendTemplateSet(templateSet ipfixentities.Set, isIPv6 bool) (int, error) {
+func (fa *flowAggregator) sendTemplateSet(isIPv6 bool) (int, error) {
 	elements := make([]*ipfixentities.InfoElementWithValue, 0)
 	ianaInfoElements := ianaInfoElementsIPv4
 	antreaInfoElements := antreaInfoElementsIPv4
@@ -486,11 +479,15 @@ func (fa *flowAggregator) sendTemplateSet(templateSet ipfixentities.Set, isIPv6 
 		ie := ipfixentities.NewInfoElementWithValue(element, nil)
 		elements = append(elements, ie)
 	}
-	err := templateSet.AddRecord(elements, templateID)
+	fa.set.ResetSet()
+	if err := fa.set.PrepareSet(ipfixentities.Template, templateID); err != nil {
+		return 0, err
+	}
+	err := fa.set.AddRecord(elements, templateID)
 	if err != nil {
 		return 0, fmt.Errorf("error when adding record to set, error: %v", err)
 	}
-	bytesSent, err := fa.exportingProcess.SendSet(templateSet)
+	bytesSent, err := fa.exportingProcess.SendSet(fa.set)
 	return bytesSent, err
 }
 

--- a/pkg/flowaggregator/flowaggregator_test.go
+++ b/pkg/flowaggregator/flowaggregator_test.go
@@ -185,12 +185,14 @@ func TestFlowAggregator_sendTemplateSet(t *testing.T) {
 			elemList = append(elemList, ipfixentities.NewInfoElementWithValue(ipfixentities.NewInfoElement(ie, 0, 0, ipfixregistry.AntreaEnterpriseID, 0), nil))
 			mockIPFIXRegistry.EXPECT().GetInfoElement(ie, ipfixregistry.AntreaEnterpriseID).Return(elemList[i+len(ianaInfoElements)+len(ianaReverseInfoElements)+len(antreaInfoElements)+len(aggregatorElements)+len(antreaSourceStatsElementList)].Element, nil)
 		}
+		mockTempSet.EXPECT().ResetSet()
+		mockTempSet.EXPECT().PrepareSet(ipfixentities.Template, testTemplateID).Return(nil)
 		mockTempSet.EXPECT().AddRecord(elemList, testTemplateID).Return(nil)
 		// Passing 0 for sentBytes as it is not used anywhere in the test. If this not a call to mock, the actual sentBytes
 		// above elements: ianaInfoElements, ianaReverseInfoElements and antreaInfoElements.
 		mockIPFIXExpProc.EXPECT().SendSet(mockTempSet).Return(0, nil)
 
-		_, err := fa.sendTemplateSet(mockTempSet, isIPv6)
+		_, err := fa.sendTemplateSet(isIPv6)
 		assert.NoErrorf(t, err, "Error in sending template record: %v, isIPv6: %v", err, isIPv6)
 	}
 }

--- a/test/e2e/flowaggregator_test.go
+++ b/test/e2e/flowaggregator_test.go
@@ -463,8 +463,9 @@ func checkRecordsForFlows(t *testing.T, data *TestData, srcIP string, dstIP stri
 
 	// Polling to make sure all the data records corresponding to the iperf flow
 	// are received.
-	err = wait.Poll(250*time.Millisecond, collectorCheckTimeout, func() (bool, error) {
-		rc, collectorOutput, _, err := provider.RunCommandOnNode(controlPlaneNodeName(), fmt.Sprintf("kubectl logs --since=%v ipfix-collector -n antrea-test", time.Since(timeStart).String()))
+	err = wait.PollImmediate(500*time.Millisecond, aggregatorInactiveFlowRecordTimeout, func() (bool, error) {
+		// `pod-running-timeout` option is added to cover scenarios where ipfix flow-collector has crashed after being deployed.
+		rc, collectorOutput, _, err := provider.RunCommandOnNode(controlPlaneNodeName(), fmt.Sprintf("kubectl logs --since=%v --pod-running-timeout=%v ipfix-collector -n antrea-test", time.Since(timeStart).String(), aggregatorInactiveFlowRecordTimeout.String()))
 		if err != nil || rc != 0 {
 			return false, err
 		}

